File: src/main/java/jugglinglab/util/Coordinate.java
//
// Coordinate.java
//
// Simple container class
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.util;

public class Coordinate {
  public double x, y, z;

  public Coordinate() {
    this(0, 0, 0);
  }

  public Coordinate(double x, double y) {
    this(x, y, 0);
  }

  public Coordinate(double x, double y, double z) {
    this.x = x;
    this.y = y;
    this.z = z;
  }

  public Coordinate(Coordinate c) {
    x = c.x;
    y = c.y;
    z = c.z;
  }

  public void setCoordinate(double x, double y, double z) {
    this.x = x;
    this.y = y;
    this.z = z;
  }

  public double getIndex(int index) {
    if (index == 0) {
      return x;
    } else if (index == 1) {
      return y;
    } else {
      return z;
    }
  }

  public void setIndex(int index, double value) {
    if (index == 0) {
      x = value;
    } else if (index == 1) {
      y = value;
    } else {
      z = value;
    }
  }

  public static Coordinate max(Coordinate coord1, Coordinate coord2) {
    if (coord1 == null) {
      return coord2;
    }
    if (coord2 == null) {
      return coord1;
    }
    return new Coordinate(
        Math.max(coord1.x, coord2.x), Math.max(coord1.y, coord2.y), Math.max(coord1.z, coord2.z));
  }

  public static Coordinate min(Coordinate coord1, Coordinate coord2) {
    if (coord1 == null) {
      return coord2;
    }
    if (coord2 == null) {
      return coord1;
    }
    return new Coordinate(
        Math.min(coord1.x, coord2.x), Math.min(coord1.y, coord2.y), Math.min(coord1.z, coord2.z));
  }

  public static Coordinate add(Coordinate coord1, Coordinate coord2) {
    if (coord1 == null) {
      return coord2;
    }
    if (coord2 == null) {
      return coord1;
    }
    return new Coordinate(coord1.x + coord2.x, coord1.y + coord2.y, coord1.z + coord2.z);
  }

  public static Coordinate sub(Coordinate coord1, Coordinate coord2) {
    if (coord1 == null) {
      return coord2;
    }
    if (coord2 == null) {
      return coord1;
    }
    return new Coordinate(coord1.x - coord2.x, coord1.y - coord2.y, coord1.z - coord2.z);
  }

  public static Coordinate truncate(Coordinate coord, double epsilon) {
    Coordinate result = new Coordinate(coord);

    if (Math.abs(result.x) < epsilon) {
      result.x = 0;
    }
    if (Math.abs(result.y) < epsilon) {
      result.y = 0;
    }
    if (Math.abs(result.z) < epsilon) {
      result.z = 0;
    }
    return result;
  }

  public static double distance(Coordinate coord1, Coordinate coord2) {
    Coordinate dc = Coordinate.sub(coord1, coord2);
    return Math.sqrt(dc.x * dc.x + dc.y * dc.y + dc.z * dc.z);
  }

  public boolean isValid() {
    return !(Double.isNaN(this.x)
        || Double.isInfinite(this.x)
        || Double.isNaN(this.y)
        || Double.isInfinite(this.y)
        || Double.isNaN(this.z)
        || Double.isInfinite(this.z));
  }

  @Override
  public String toString() {
    return "("
        + JLFunc.toStringRounded(x, 5)
        + ", "
        + JLFunc.toStringRounded(y, 5)
        + ", "
        + JLFunc.toStringRounded(z, 5)
        + ")";
  }
}

File: src/main/kotlin/jugglinglab/util/Coordinate.kt
//
// Coordinate.kt
//
// Defines an (x, y, z) spatial coordinate.
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.util

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

data class Coordinate @JvmOverloads constructor(
    @JvmField var x: Double = 0.0,
    @JvmField var y: Double = 0.0,
    @JvmField var z: Double = 0.0
) {
    fun setCoordinate(x: Double, y: Double, z: Double) {
        this.x = x
        this.y = y
        this.z = z
    }

    fun get(index: Int): Double {
        return when (index) {
            0 -> x
            1 -> y
            2 -> z
            else -> throw IndexOutOfBoundsException("Invalid Coordinate index: $index")
        }
    }

    fun set(index: Int, value: Double) {
        when (index) {
            0 -> x = value
            1 -> y = value
            2 -> z = value
            else -> throw IndexOutOfBoundsException("Invalid Coordinate index: $index")
        }
    }

    val isValid: Boolean
        get() = !(x.isNaN()
            || x.isInfinite()
            || y.isNaN()
            || y.isInfinite()
            || z.isNaN()
            || z.isInfinite())

    override fun toString(): String {
        return ("("
            + toStringRounded(x, 5)
            + ", "
            + toStringRounded(y, 5)
            + ", "
            + toStringRounded(z, 5)
            + ")")
    }

    companion object {
        fun max(coord1: Coordinate?, coord2: Coordinate?): Coordinate? {
            if (coord1 == null) {
                return coord2
            }
            if (coord2 == null) {
                return coord1
            }
            return Coordinate(
                max(coord1.x, coord2.x), max(coord1.y, coord2.y), max(coord1.z, coord2.z)
            )
        }

        fun min(coord1: Coordinate?, coord2: Coordinate?): Coordinate? {
            if (coord1 == null) {
                return coord2
            }
            if (coord2 == null) {
                return coord1
            }
            return Coordinate(
                min(coord1.x, coord2.x), min(coord1.y, coord2.y), min(coord1.z, coord2.z)
            )
        }

        fun add(coord1: Coordinate, coord2: Coordinate): Coordinate {
            return Coordinate(coord1.x + coord2.x, coord1.y + coord2.y, coord1.z + coord2.z)
        }

        fun sub(coord1: Coordinate, coord2: Coordinate): Coordinate {
            return Coordinate(coord1.x - coord2.x, coord1.y - coord2.y, coord1.z - coord2.z)
        }

        fun truncate(coord: Coordinate, epsilon: Double): Coordinate {
            val result: Coordinate = coord.copy()

            if (abs(result.x) < epsilon) {
                result.x = 0.0
            }
            if (abs(result.y) < epsilon) {
                result.y = 0.0
            }
            if (abs(result.z) < epsilon) {
                result.z = 0.0
            }
            return result
        }

        fun distance(coord1: Coordinate, coord2: Coordinate): Double {
            val dc: Coordinate = sub(coord1, coord2)
            return sqrt(dc.x * dc.x + dc.y * dc.y + dc.z * dc.z)
        }
    }
}
